File: C++/Curso Estruct. Abs. De Datos y Algoritmos/Laboratorios/Lab3GeometriaPython/sourcecode/Imports/Figura.py
#@file Figura.py
from abc import *
from .Vertice import Vertice


#@brief Clase que crea Figuras en el espacio 2D
class Figura(Vertice):
    universalid=0
    #@brief constructor de figura
    def __init__(self,vertices):
        
        
        self.nombre=""
        self.color=""
        self.paraImprimir = ""; 
        self.identificador=Figura.universalid
        Figura.universalid=Figura.universalid+1
        self.listaVertices = []
        self.lado=0
        self.area=0
        self.perimetrofig=0
        self.cantidadPuntos = 0
        self.superficie()
        self.perimetro()
        
    #@brief calcula el area de una figura por defecto
    @abstractmethod
    def superficie(self):
        self.area=self.lado**2
        return (self.area)

    #@brief calcula el perimetro de una figura por defecto
    @abstractmethod
    def perimetro(self):
        self.perimetrofig=self.lado*(self.cantidadPuntos)
        return (self.perimetro)
    
    #@brief Sobrecarga del operador ~ 
    #@return Devuelve la representacion en texto de una Figura
    def __invert__(self):
        respuesta = "Lista de vertices \n"
        for i in range(0,self.cantidadPuntos):
            respuesta = respuesta + "Vertice " + str(i+1) + "\n"
            respuesta = respuesta + ~self.listaVertices[i] + "\n"
        respuesta = respuesta + "\n"
        respuesta = respuesta + "Mi nombre es: " + str(self.nombre) + "\n"
        respuesta = respuesta + "Mi color es: " + str(self.color) + "\n"
        respuesta = respuesta + "Mi superfice es: " + str(self.area) + "\n"
        respuesta = respuesta + "Mi perimetro es: " + str(self.perimetrofig) + "\n"
        self.paraImprimir = self.paraImprimir + "\n" + respuesta
        return respuesta
    

File: C++/Curso Estruct. Abs. De Datos y Algoritmos/Laboratorios/lab2 Geometria/sourcecode/Figura.cpp
/**
 *@file Figura.cpp/
 */
#include "../include/Includes.hpp"

using namespace std;

/**
 * @brief constructor por defecto
 */
Figura::Figura(){
    //this->identificadorEstatico = this->identificadorEstatico + 1;
}
/**
 * @brief destructor por defecto
 */
Figura::~Figura(){
}
/**
 * @brief constructor por copia
 * @param f es un objeto
 */
Figura::Figura(const Figura& f){
    this->area = f.area;
    this->perimetrofig = f.perimetrofig;
    this->nombre = f.nombre;
    this->color = f.color;
    this->identificadorEstatico = f.identificadorEstatico;

}

/**
 * @brief metodo que calcula una superficie generica
 */
double Figura::superficie(){
    double superficie = this->x * this->y;
    return superficie;
}

/**
 * @brief metodo que calcula un perimetro generico
 */
double Figura::perimetro(){
    double perimetro = 4*x;
    return perimetro;
}

/**
 * @brief sobrecarga del operador ~ 
 */
string Figura::operator~(){
    string retorno;
    string strpuntos = "";
    for (int i=0; i < this->cantidadPuntos; i++){
        strpuntos = strpuntos + ~puntos[i];
        strpuntos = strpuntos + "\n";
    }
    retorno = strpuntos + "\n";
    retorno = retorno + ("Mi nombre es "+this->nombre + "  \n" +
               "Mi color es "+this->color + " \n" +
               "Mi superficie es "+to_string(this->superficie())+ " \n" +
               "Mi perimetro es "+to_string(this->perimetro())+ " \n"
               );
    return retorno;
};
